### PR TITLE
feat(chart): enable device auth grant

### PR DESCRIPTION
**Pull request by mohammad-sdsc**
Fixes: #2490

### CODE REVIEW

**Review by ableuler — COMMENTED** [review]
How do you want to handle the transition period? I guess one option would be for the server side to support both, old and new CLI login for a certain period.

**Comment by mohammad-sdsc** [comment]
Yes, that would work. This PR does not remove anything so, it's safe to merge it. We should keep the custom auth flow in the gateway for a while (i.e. not merging https://github.com/SwissDataScienceCenter/renku-gateway/pull/558 for now) to support old CLI clients.

**Comment by ableuler** [comment]
Sounds good. But for this to work we'd have to still allow the standard flow on the renku-cli for the time being, no?

**Comment by mohammad-sdsc** [comment]
I've checked the `init-realm.py` script and it doesn't update clients automatically. It only warns about mismatches between existing and new clients. I see that this scripts hasn't been updated in ages so I'm not sure what's the policy regarding it: Should we update it or should we apply changes to the deployments manually. If former, then I can add a section to the script to copy new client attributes to an existing one.

**Comment by ableuler** [comment]
Thanks for checking. I think the post-install job should indeed delete and recreate clients which do not match the expected configuration.

**Comment by rokroskar** [comment]
Is this waiting for a change to the post-install job?

**Comment by rokroskar** [comment]
@mohammad-sdsc thanks for looking into this! Can't we just nuke the old configuration and apply the new one in the case that there is a config change? That's effectively what we do by hand now and it works fine...

**Comment by mohammad-sdsc** [comment]
> Can't we just completely remove the old configuration and apply the new one in the case that there is a config change? That's effectively what we do by hand now and it works fine...

I didn't realize this. Thanks for mentioning it! It's more robust and simpler to implement.

**Comment by rokroskar** [comment]
@ableuler please chime in if I misrepresented how this is currently handled :)

**Comment by ableuler** [comment]
> Can't we just completely remove the old configuration
What do you mean by existing configuration? Do you mean _all_ clients in the Renku realm?

**Comment by rokroskar** [comment]
I meant the client that has the configuration conflict. Though in practice we actually removed all the clients and let the post-install script recreate them iirc, but I wouldn't do that in general.

**Comment by mohammad-sdsc** [comment]
> I never delete all clients by hand and even more so I think that we should not do this programmatically.

It doesn't delete all clients; it only deletes/recreates clients that have changed. The difference with the previous approach is in the configuration that is used to recreate the client: Before it was using the existing configuration but now it's using the new configuration. Basically, this should be fine if all the required configuration is already in the value files/templates. If there are manual changes to the clients (i.e. changes via Keycloak UI) then those are gone.

**Comment by ableuler** [comment]
Yeah, the `recreate client instead of update` aspect makes perfect sense to me. I just wasn't sure if Rok was talking about the same thing. Anyway, I think we all agree.

File: helm-chart/renku/templates/post-install-job-keycloak.yaml [hunk 1]
```diff
@@ -33,7 +33,8 @@ spec:
             "--admin-password=$(KEYCLOAK_ADMIN_PASSWORD)",
             "--keycloak-url=$(KEYCLOAK_URL)",
             "--users-file=/app/data/users",
-            "--clients-file=/app/data/clients"
+            "--clients-file=/app/data/clients",
+            "--apply-changes"
```
Comment (ableuler) [inline helm-chart/renku/templates/post-install-job-keycloak.yaml:37]: maybe after your most recent edit calling this just `--force` might be more descriptive?

**Comment by mohammad-sdsc** [comment]
Follow up story: https://github.com/SwissDataScienceCenter/renku/issues/2504

**Review by ableuler — APPROVED** [review]
Thanks @mohammad-sdsc, let's get this merged!

**Pull request merged**